File: go/weak_cryptography/random_tokens.go
package weakcryptography

import (
	"fmt"
	"math/rand"
)

// VULN 1: math/rand used for password reset tokens - predictable token generation
func GeneratePasswordResetToken() string {
	token := rand.Intn(999999999)
	return fmt.Sprintf("%09d", token)
}

// VULN 2: math/rand Int63 used for order confirmation codes - guessable codes
func GenerateOrderConfirmationCode() int64 {
	return rand.Int63()
}

// VULN 3: math/rand seeded with userID for CSRF tokens - user-predictable values
func GenerateCSRFToken(userID int64) string {
	r := rand.New(rand.NewSource(userID))
	return fmt.Sprintf("%016x", r.Int63())
}

File: python/weak_cryptography/random_tokens.py
import random
import string
import math


# VULN 1: random.random() used to generate password reset tokens - not cryptographically secure
def generate_password_reset_token():
    token = ""
    for _ in range(32):
        token += random.choice(string.ascii_letters + string.digits)
    return token


# VULN 2: random.randint() used for order confirmation codes
def generate_order_confirmation_code():
    return random.randint(100000, 999999)


# VULN 3: math.random equivalent (random module) for CSRF token generation
def generate_csrf_token(user_id):
    random.seed(user_id)
    return "".join([str(random.randint(0, 9)) for _ in range(16)])
